### ISSUE TITLE
2.0.0a2 L003 triggered within MERGE resulting in MERGE on a new line

### ISSUE THREAD

**Issue by fmms** [sqlfluff/sqlfluff#4250]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

We have lots of I have lots of complex ``MERGE` statements in my code base but let's start at the one from the Microsoft documentation (see https://learn.microsoft.com/en-us/sql/t-sql/statements/merge-transact-sql?view=sql-server-ver16)

```
-- Create a temporary table variable to hold the output actions.  
DECLARE @SummaryOfChanges TABLE(Change VARCHAR(20));  

MERGE INTO Sales.SalesReason AS tgt  
USING (VALUES ('Recommendation','Other'), ('Review', 'Marketing'),
              ('Internet', 'Promotion'))  
       as src (NewName, NewReasonType)  
ON tgt.Name = src.NewName  
WHEN MATCHED THEN  
UPDATE SET ReasonType = src.NewReasonType  
WHEN NOT MATCHED BY TARGET THEN  
INSERT (Name, ReasonType) VALUES (NewName, NewReasonType)  
OUTPUT $action INTO @SummaryOfChanges;  
```

When formatting this, the ``THEN`s are move to a separate line which is a mistake in my opinion.

Resulting in
```
-- Create a temporary table variable to hold the output actions.  
DECLARE @SummaryOfChanges TABLE(Change varchar(20));

MERGE INTO Sales.SalesReason tgt
USING (
  VALUES ('Recommendation', 'Other'), ('Review', 'Marketing')
  , ('Internet', 'Promotion')
)
src (NewName, NewReasonType)
  ON tgt.Name = src.NewName
WHEN MATCHED
  THEN
  UPDATE SET ReasonType = src.NewReasonType
WHEN NOT MATCHED BY TARGET
  THEN
  INSERT (Name, ReasonType) VALUES (NewName, NewReasonType)
OUTPUT $action INTO @SummaryOfChanges;
```

### Expected Behaviour

The ``WHEN NOT MATCHED THEN``` kind of clauses stay on a single line.

### Observed Behaviour

``THEN` is moved to another line

### How to reproduce

Take above query and run `sqlfluff fix .\merge.sql  --dialect tsql`

### Dialect

tsql

### Version

```
==== sqlfluff ====
sqlfluff:              2.0.0a2 python:                 3.11.0
implementation:        cpython verbosity:                   2
dialect:                  tsql templater:                 raw
rules:                                all
== Raw Config:
core:
```

### Configuration

none

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by alanmcruickshank** [comment]
Yep - looking at the current dialect implementation, the `Indent` is set to come _before_ the `THEN` in a `MERGE` statement. Currently L003 doesn't _remove_ unnecessary indent, but this formulation of the query should pass:

```tsql
-- Create a temporary table variable to hold the output actions.  
DECLARE @SummaryOfChanges TABLE(Change varchar(20));

MERGE INTO Sales.SalesReason tgt
USING (
  VALUES ('Recommendation', 'Other'), ('Review', 'Marketing')
  , ('Internet', 'Promotion')
)
src (NewName, NewReasonType)
  ON tgt.Name = src.NewName
WHEN MATCHED
  THEN UPDATE SET ReasonType = src.NewReasonType
WHEN NOT MATCHED BY TARGET
  THEN INSERT (Name, ReasonType) VALUES (NewName, NewReasonType)
OUTPUT $action INTO @SummaryOfChanges;
```

On the question of whether we should shift the location of the indent, I'm open to it but I'm not a regular TSQL user, and reading the docs you link to it looks like Microsoft are a little inconsistent in their use. The original implementation was in #2057. @tunetheweb, @jpers36, @tkachenkomaria244 do you folks have views on this as part of that implementation?

For my 2c reading the docs, it does look like indenting _after_ the `THEN` is more common _unless there is an `AND` expression_, in which case it's more normal to indent _before_ the `THEN`. This shift in behaviour does feel a little strange, and I can understand why if the expression gets long that it would be better to have the indent before the `THEN` so you can find it. That's making me think that the current implementation does make sense. In any case, whatever we do with the default, this should be easily configurable.

**Comment by tunetheweb** [comment]
I would have thought the THEN is part of then WHEN clause so to me it makes to indent before the THEN.

**Comment by tunetheweb** [comment]
However I can see the sense to also indent after the THEN if that sub-clause goes to multiple lines.

**Comment by alanmcruickshank** [comment]
I've removed myself as the assignee for this one - because I don't know if I'm the best person to fix it, and also because I don't think it's on the critical path for 2.0.0 release.

If someone does want to pick this up, I currently think it's best solved by adding a configurable indent flag for merge statements (as per `indented_joins` described in the docs here: https://docs.sqlfluff.com/en/stable/layout.html#configuring-indent-locations). That would allow users who want their indent to come after the `THEN` keyword and not _before_ have a configuration flag to set that.

I'm also very open to more creative solutions to this issue - but given I'm not a TSQL expert, I think I'm a little out of my depth on how best to devise something this is consistent across a range of scenarios.
